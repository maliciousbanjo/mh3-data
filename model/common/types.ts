/** Status ailment types */
export enum Status {
  POISON = "poison",
  PARALYSIS = "paralysis",
  SLEEP = "sleep",
  STUN = "stun",
  EXHAUST = "exhaust",
}

/** Element types */
export type Element = "FIRE" | "WATER" | "THUNDER" | "ICE" | "DRAGON";

/** Levels of item/equipment rarity */
export type Rarity = 1 | 2 | 3 | 4 | 4 | 5 | 6 | 7;

/** Possible number of decoration slots on armors and weapons */
export type Slots = 0 | 1 | 2 | 3;