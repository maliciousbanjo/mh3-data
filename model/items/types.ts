/**
 * Ingame item eg; Mega Potion, Raw Meat, etc
 */
export interface Item {
  readonly id: number;
  readonly name: string;
}

export type ItemRecord = Record<Item["id"], Item>;

/**
 * Result of gathering/carving/quest reward
 *
 * Key:Value pair of an {@link Item} id and a percentage chance.
 */
export interface ItemResult {
  itemId: Item['id'];
  /** Percentage */
  chance: number;
}

/**
 * Quest reward item
 */
export interface QuestReward {
  itemId: Item['id'];
  /** Percentage */
  chance: number;
  /** Amount received */
  count: number;
}

/**
 * Amount needed of a particular item for crafting, trading, etc
 */
export interface ItemRequirement {
  itemId: Item['id'];
  amount: number;
}