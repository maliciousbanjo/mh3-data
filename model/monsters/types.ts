import { Status } from "../common";
import { QuestReward, ItemResult } from "../items";

export enum MonsterType {
  Herbvr = "herbivore",
  Nptron = "neopteron",
  Lynian = "lynian",
  Fish = "fish",
  PscWyv = "piscine wyvern",
  BrdWyv = "bird wyvern",
  FlyWyv = "flying wyvern",
  Levthn = "leviathan",
  BrtWyv = "brute wyvern",
  EldDrg = "elder dragon",
}

/**
 * Monster hitzone values for different damage types.
 * Used by {@link Hitzone} to match against a particular monster part
 */
export interface HitzoneValues {
  cut: number;
  impact: number;
  gun: number;
  fire: number;
  water: number;
  thunder: number;
  ice: number;
  dragon: number;
  stagger: number;
}

/**
 * Record of <Part: HitzoneValues>
 */
export type Hitzone = Record<string, HitzoneValues>;

/**
 * Collection of either a default hitzone or hitzones
 * based on a monster's status eg; enraged, flying, etc
 */
export interface HitzoneGroup {
  name: string;
  hitzones: Hitzone;
}

/**
 * Monster part that can be broken
 */
export interface Breakable {
  name: string;
  /** Low rank break rewards */
  low?: QuestReward[];
  /** High rank break rewards */
  high?: QuestReward[];
}

/**
 * Criteria and rewards for a monster being captured
 */
export interface CaptureInfo {
  /** Health percentage threshold for a monster to be capture-ready */
  health: number;
  /** Low rank capture rewards */
  low?: ItemResult[];
  /** High rank capture rewards */
  high?: ItemResult[];
}

/**
 * Status information for a particular {@link Status} ailment
 */
export interface StatusValues {
  /** Duration of status effect in seconds */
  duration: number;
  /** Resistance/buildup of status damage until effect activates */
  tolerance: {
    /** Value required for the first status effect */
    initial: number;
    /** Maximum resistance */
    max: number;
  };
  /** Amount of status damage recovered per interval */
  recovery: {
    /** Damage recovered */
    amount: number;
    /** Interval of recovery in seconds */
    wait: number;
  };
}

/**
 * Carvable monster part
 */
export interface CarveZone {
  name: string;
  /** Number of possible carves */
  count: number;
  /** Low rank carve */
  low?: ItemResult[];
  /** High rank carve */
  high?: ItemResult[];
}

/**
 * Extra unique data for a particular monster
 */
export interface Variant {
  name: string;
  carves?: CarveZone[];
  shiny?: ShinyDrop;
}

/**
 * Item result for a monster shiny drop
 */
export interface ShinyDrop {
  /** Action required to cause the shiny to drop */
  action: string;
  /** Low rank result */
  low?: ItemResult[];
  /** High rank result */
  high?: ItemResult[];
}

/**
 * Base properties for any monster
 */
export interface Monster {
  id: number;
  name: string;
  description: string;
  type: MonsterType;
  threat: number;
  /** 
   * List of hitzones. Will either contain a "default" hitzone group 
   * or multiple based on different monster states eg; flying, enraged, etc
   */
  hitzoneGroups: HitzoneGroup[];
  /** If undefined this monster should have a 'variant' property */
  shiny?: ShinyDrop;
  /** If undefined this monster should have a 'variant' property */
  carves?: CarveZone[];
  /** Optional support for varying shines or carves */
  variants?: Variant[];
}

/**
 * Properties for a large (boss) monster
 */
export interface LargeMonster
  extends Omit<Monster, "carves" | "type" | "variants"> {
  hp: number;
  type:
    | MonsterType.BrdWyv
    | MonsterType.FlyWyv
    | MonsterType.Levthn
    | MonsterType.BrtWyv
    | MonsterType.EldDrg;
  /** Quests that include this monster */
  quests: {
    // TODO: quest type. Will be closely tied to the quest data.
  };
  carves: CarveZone[];
  breakables: Breakable[];
  status?: {
    [Status.POISON]: StatusValues & {
      /** Damage taken by poison per interval */
      damage: {
        initial: number;
        total: number;
      };
    };
    [Status.PARALYSIS]: StatusValues;
    [Status.SLEEP]: StatusValues;
    [Status.STUN]: StatusValues;
    [Status.EXHAUST]?: Omit<StatusValues, "duration">;
  };
  /** {@link MonsterType.EldDrg} cannot be captured */
  capture?: CaptureInfo;
}

export type MonsterRecord = Record<Monster["id"], Monster>;
export type LargeMonsterRecord = Record<LargeMonster["id"], LargeMonster>;
