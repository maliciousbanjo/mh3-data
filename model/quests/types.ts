import { Item, QuestReward } from "../items";
import { Monster } from "../monsters";
import { MonsterLevel } from "../monster-levels";

export type QuestType =
  | "gather"
  | "hunt"
  | "capture"
  | "harvest"
  | "escort"
  | "multihunt"
  | "repel"
  | "slay";

export type QuestGoal =
  | "gathering"
  | "slay"
  | "capture"
  | "hunt"
  | "break"
  | "sever"
  | "sonic"
  | "repel"
  | "other"
  | "anthill";

export enum Location {
  DESERTED_ISLAND = "Deserted Island",
  SANDY_PLAINS = "Sandy Plains",
  FLOODED_FOREST = "Flooded Forest",
  TUNDRA = "Tundra",
  VOLCANO = "Volcano",
  UNDERWATER_RUIN = "Underwater Ruin",
  GREAT_DESERT = "Great Desert",
  SACRED_LAND = "Sacred Land",
  LAND_ARENA = "Land Arena",
  WATER_ARENA = "Water Arena",
}

/**
 * These properties were derived from the publicly-available MH3 Quest editor.
 * The values are incorrect/flawed. ZeSpyro
 */
interface BossInfo {
  type1: Monster["id"];
  /** Starting area? */
  type2: 0 | 1 | 2;
  /** TODO: Unidentified monster levels: 30, 55, 59, 60 */
  level: MonsterLevel | 30 | 55 | 59 | 60;

  size: number;
  /** controls random distribution of monster's HP */
  min: number;
  /** controls random distribution of monster's size */
  max: number;
}

/**
 * Base quest type
 *
 * It looks like different questTypes have different properties
 * 'target' points to differnt ids. Gather quests have IDs for items. Hunt/capture likely
 * have IDs for monsters, but the monsters themselves don't seem to have IDs.
 */
interface Quest {
  name: string;
  client: string;
  type: QuestType;
  goal: QuestGoal;
  /** Cost to initiate quest in zenny */
  contract: number;
  /** Maximum quest time in minutes */
  time: number;
  location: Location;
  /** If true, hunter will spawn in random location */
  randomspawn: boolean;
  /** Quest completion reward in zenny */
  reward: number;
  /** Amount of zenny taken from reward per faint */
  ko: number;
  /** Hunter Rank Points received for quest completion */
  hrp: number;
  secondaryRewards: QuestReward[];
  /** Key will be an ID for an item or monster depending on the {@link QuestType} */
  targets: Record<number, number>;
  /** Large monsters that can spawn in this quest */
  bosses: BossInfo[];
  // TODO: Will need to create different subquests types with different target properties
  subquests?: object;
}

type NumberRequired = number;

export interface GatheringQuest extends Quest {
  type: "gather";
  goal: "gathering";
  targets: Record<Item["id"], NumberRequired>;
}

export interface HarvestQuest extends Quest {
  type: "harvest";
  goal: "gathering";
  /** Harvest Ticket */
  targets: {
    396: 1;
  };
}

export interface HuntQuest extends Quest {
  type: "hunt";
  goal: "hunt" | "slay";
  targets: Record<Monster["id"], NumberRequired>;
  bosses: BossInfo[];
}

export interface MultiHuntQuest extends Quest {
  type: "multihunt";
  goal: "hunt";
  targets: Record<Monster["id"], NumberRequired>;
}

export interface CaptureQuest extends Quest {
  type: "capture";
  goal: "capture";
  targets: Record<Monster["id"], NumberRequired>;
  bosses: BossInfo[];
}

export interface EscortQuest extends Quest {
  type: "escort";
  goal: "slay";
  targets: Record<Monster["id"], NumberRequired>;
}

export interface ArenaQuest extends Omit<Quest, "secondaryRewards"> {
  type: "slay";
  goal: "hunt";
  targets: Record<Monster["id"], NumberRequired>;
  bosses: BossInfo[];
}

export interface RepelQuest extends Quest {
  type: "repel";
  goal: "repel" | "other";
  bosses: BossInfo[];
}

/** Quest star levels */
export enum StarLevel {
  OneStar = "1 Star",
  TwoStar = "2 Star",
  ThreeStar = "3 Star",
  FourStar = "4 Star",
  FiveStar = "5 Star",
  SixStar = "6 Star",
}

export interface QuestRegion {
  name: "Village" | "City";
  starLevels: {
    [StarLevel.OneStar]: Quest[];
    [StarLevel.TwoStar]: Quest[];
    [StarLevel.ThreeStar]: Quest[];
    [StarLevel.FourStar]: Quest[];
    [StarLevel.FiveStar]: Quest[];
    [StarLevel.SixStar]: Quest[];
  };
  arena: ArenaQuest[];
  events: Quest[];
}
