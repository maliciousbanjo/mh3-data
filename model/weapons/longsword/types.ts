import { Weapon, WeaponType } from "../types";

export type LongswordAttack =
  | "Unsheathe Attack"
  | "Step Stash"
  | "Downward Slash"
  | "Stab"
  | "Upward Slash"
  | "Sweep"
  | "Empty Spirit"
  | "Spirit 1"
  | "Spirit 2"
  | "Spirit 3"
  | "Spirit Finisher";

export type Longsword = Weapon<WeaponType.LONGSWORD>;
