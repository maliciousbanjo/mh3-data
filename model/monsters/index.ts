export * from "./types";
export { LargeMonsterData } from "./large-monster-data";
export { SmallMonsterData } from "./small-monster-data";
