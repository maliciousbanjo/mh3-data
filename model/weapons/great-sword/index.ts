export * as GreatSwordTypes from './types'
export { GreatSwords, GreatSwordDamageProperties } from './great-sword-data'