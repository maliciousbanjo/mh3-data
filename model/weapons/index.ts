// Data
import { GreatSwords, GreatSwordDamageProperties } from "./great-sword";
import {
  SwordAndShields,
  SwordAndShieldDamageProperties,
} from "./sword-and-shield";
import { Hammers, HammerDamageProperties } from "./hammer";
import { Lances, LanceDamageProperties} from './lance';

// Types
export type { GreatSwordTypes } from "./great-sword";
export type { SwordAndShieldTypes } from "./sword-and-shield";
export type { HammerTypes } from "./hammer";
export type { LanceTypes } from './lance';
export * from "./types";

export { GreatSwords, SwordAndShields, Hammers, Lances };

export const DamageProperties = {
  GreatSwordDamageProperties,
  SwordAndShieldDamageProperties,
  HammerDamageProperties,
  LanceDamageProperties
};
