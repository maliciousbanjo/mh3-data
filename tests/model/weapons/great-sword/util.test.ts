import { Alatreon } from '../../../../src/model/monsters/large-monster-data';
import { GreatSwords } from '../../../../src/model/weapons/great-sword';
import { calculateGreatSwordDamage } from '../../../../src/model/weapons/great-sword/util';
import { Sharpness } from '../../../../src/model/weapons/types';
import { MonsterLevels } from '../../../../src/model';

describe('Great Sword utils', () => {
  describe('calculateGreatSwordDamage', () => {
    it('calculates damage', () => {
      const anguishP = GreatSwords[57];
      const headHitzoneValues = Alatreon.hitzoneGroups[0].hitzones['Head']; // Non-flying head hitzone
      const levelMultipliers =
        MonsterLevels.Util.getMonsterLevelMultipliers(52); // The Brilliant Darkness
      const result = calculateGreatSwordDamage(
        anguishP,
        'L3 Charge',
        Sharpness.PURPLE,
        headHitzoneValues,
        levelMultipliers
      );
      console.log(`${anguishP.name}, purple, L3 Charge, Alatreon Head`, result);
      expect(result).toMatchSnapshot();
    });
  });
});
