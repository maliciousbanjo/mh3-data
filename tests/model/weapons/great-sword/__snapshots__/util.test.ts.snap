// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Great Sword utils calculateGreatSwordDamage calculates damage 1`] = `
[
  {
    "elementalDamage": 0.675,
    "koDamage": undefined,
    "rawDamage": 144.78750000000002,
    "totalDamage": 145,
  },
]
`;
